### DISCUSSION TITLE
When using register_block_type is there a way to manually declare a dependency?

### DISCUSSION THREAD

**Discussion by jakeparis** [WordPress/gutenberg#46589] [Q&A]
I have a block which requires a script from one of our mu-plugins. However, when I register the block using `register_block_type` and a block.json file, there's no way to declare a dependency. 

I'm looking to declare a dependency in the same way I would as the third parameter in `wp_enqueue_script()`

**Comment by t-hamano** [comment]
@jakeparis 
I think we can specify script and style dependencies by handles in `block.json`. Would this be a solution?

https://developer.wordpress.org/block-editor/reference-guides/block-api/block-metadata/#assets

**Reply by jakeparis** [reply]
@t-hamano Maybe. But it looks like, in block.json, there are two ways to specify an asset... with the **WPDefinedPath** shape: 

```json
{ 
    "editorScript": [ "file:./index.js" ]
}
```

or with the **WPDefinedAsset** shape, which seems to be only for scripts which are already registered elsewhere:

```json
{
    "editorScript": {
         "dependencies": [
            "other-script",
            "third-script",
         ],    
    }
}
```

But how would I use the "file:./index.js" style to register the script, and also declare the dependencies at the same time. Is there a way to do that?

**Reply by t-hamano** [reply]
In the navigation block, two file paths are defined in `viewScript`.

https://github.com/WordPress/gutenberg/blob/d9f13a8d7c6930b974022d90d6f1dc6c16a13f0f/packages/block-library/src/navigation/block.json#L136

I have not tried this, but could this be applied to `editorScript` as well?

**Reply by jakeparis** [reply]
defining two files paths like that is to register both of them. That is not a method to declare a dependency.

**Reply by t-hamano** [reply]
How about determining if the dependent plugins are enabled and registering the block?
I haven't tested it, but it probably looks like this:

```PHP
if ( is_plugin_active( '/path/to/plugin.php' ) {
  register_block_type(...);
}
```

**Reply by jakeparis** [reply]
No, unfortunately that's a totally different thing. Knowing whether a plugin is active is different than being sure a script is enqueued before my block/script.

**Comment by gziolo** [comment]
As of today, the script needs to be registered with `register_block_type` and then can be connected with the block using the script handle:

```json
{ 
    "editorScript": [ "file:./index.js", "my-script-with-custom-dependencies" ]
}
```

However, we would be happy to see implemented the proposed solution in https://github.com/WordPress/gutenberg/issues/46954.
